src/daft-local-execution/src/sources/scan_task.rs

```diff
@@ -3,115 +3,121 @@ use std::{
+                    }
+                }
+                count
+            }
+            // If there's no row limit, use all available CPU cores
+            None => *NUM_CPUS,
+        };
+        Self {
+            scan_tasks,
+            num_parallel_tasks,
+            schema,
         }
-        Ok(())
     }
-    pub fn boxed(self) -> Box<dyn Source> {
-        Box::new(self) as Box<dyn Source>
+
+    pub fn arced(self) -> Arc<dyn Source> {
+        Arc::new(self) as Arc<dyn Source>
     }
 }
+
+#[async_trait]
 impl Source for ScanTaskSource {
     #[instrument(name = "ScanTaskSource::get_data", level = "info", skip_all)]
-    fn get_data(
+    async fn get_data(
         &self,
         maintain_order: bool,
-        runtime_handle: &mut ExecutionRuntimeHandle,
         io_stats: IOStatsRef,
-    ) -> crate::Result<SourceStream<'static>> {
-        let (senders, receivers): (Vec<_>, Vec<_>) = if maintain_order {
-            (0..self.scan_tasks.len())
-                .map(|_| create_channel(1))
-                .unzip()
-        } else {
-            let (sender, receiver) = create_channel(self.scan_tasks.len());
-            (
-                std::iter::repeat(sender)
-                    .take(self.scan_tasks.len())
-                    .collect(),
-                vec![receiver],
-            )
-        };
-        let scan_tasks = self.scan_tasks.clone();
-        runtime_handle.spawn(
-            async move {
-                let mut task_set = TaskSet::new();
-                let delete_map = get_delete_map(&scan_tasks).await?.map(Arc::new);
-                for (scan_task, sender) in scan_tasks.into_iter().zip(senders) {
-                    task_set.spawn(Self::process_scan_task_stream(
-                        scan_task,
-                        sender,
-                        maintain_order,
+    ) -> DaftResult<SourceStream<'static>> {
+        let io_runtime = get_io_runtime(true);
+        let delete_map = get_delete_map(&self.scan_tasks).await?;
+        let stream =
+            futures::stream::iter(self.scan_tasks.clone().into_iter().map(move |scan_task| {
+                let io_stats = io_stats.clone();
+                let delete_map = delete_map.clone();
+                io_runtime.spawn(async move {
+                    stream_scan_task(
+                        scan_task.clone(),
                         io_stats.clone(),
-                        delete_map.clone(),
-                    ));
-                }
-                while let Some(result) = task_set.join_next().await {
-                    result.context(JoinSnafu)??;
-                }
-                Ok(())
-            },
-            self.name(),
-        );
+                        delete_map.map(Arc::new),
+                        maintain_order,
+                    )
+                    .await
+                })
+            }))
+            .buffered(self.num_parallel_tasks)
+            .map(|s| s?)
+            .try_flatten();
```

Comment:
This is actually a pretty important fix because previously all the scan tasks were spawned on the execution_runtime, which is single threaded. I tested TPCH with num_files = 2 x num_cpus, and this change makes it much faster. On that note, we should make our benchmarks test not just 1 / 2 files, but also > num_cpu files.